keps/sig-node/3085-pod-conditions-for-starting-completition-of-sandbox-creation/README.md

```diff
@@ -0,0 +1,1392 @@
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+None so far
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+SLOs are not applicable to pod status fields. Overall Kubernetes node level SLOs
+may leverage this feature.
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+The main drawback associated with the new pod sandbox conditions involves a
+slight potential increase in calls to the API Server from Kubelet to patch
+`status` = `true` for the new conditions in a pod's status. Typically, this
+would involve two extra patch calls for pod status in the lifetime of most pods
+(if the status manager does not batch them with other pod status updates): one
+when pod sandbox creation actually starts and the other when pod sandbox
+creation completes successfully. However, there could be a higher number of
+patch calls to API Server if the pod sandbox environment (like a microvm) starts
+successfully and then crashes in a re-start loop.
+
+Caching of updates to pod status by the pod status manager and batching pod
+status updates (which is already in place) can help mitigate frequent patch
+calls to API server.
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+### Dedicated fields or annotations for the pod sandbox creation timestamps
+
+Timestamps around start and completion of pod sandbox creation may be surfaced
+as dedicated fields in the pod status rather than pod conditions. However, since
+the start and creation of pod sandbox are essentially "milestones" in the life
+of a pod (similar to Scheduled, Ready, etc), pod conditions is the ideal place
+to surface these.
+
+Dedicated annotations on the pod for surfacing this data is another potential
+approach. However, usage of annotations for Kubelet managed data is typically
+discouraged.
+
+### Surface pod sandbox creation latency instead of timestamps
+
+Surfacing the amount of time it took to successfully create a pod sandbox is an
+alternative to surfacing the timestamps around start and completion of pod
+sandbox creation (whose delta reflects the latency). The latency data would
```

Comment:
Do we care about latency info for each individual pod, or in aggregate? If the former, do traces suffice? If the latter, would a metric?

Reply:
For data collection use-cases, we want the sandbox creation conditions to be surfaced on a per-pod cardinality with full availability to the fields of the pod (and related objects like PVCs, in-line volumes, etc) that a service can use to aggregate the data across (a) arbitrary labels and annotations in a pod denoting classes of workloads (b) fields like storage classes of PVCs mounted by the pod, runtimeclass, etc. without any sampling. 

For controller use-cases (that manages pods and associated objects like PVCs/CRDs associated with networking, etc), the data as sandbox conditions (or dedicated pod fields) would be ideal while traces would be sub-optimal.

Reply:
Clarified [here](https://github.com/kubernetes/enhancements/pull/3087/files#diff-08c9a02aebe0409f9d02d84d2a337e642394319991f684d82798f400510a6d85R1509) and [here](https://github.com/kubernetes/enhancements/pull/3087/files#diff-08c9a02aebe0409f9d02d84d2a337e642394319991f684d82798f400510a6d85R1518)